moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,553 @@
+ * ```
+ *
+ * Each `JsonWriter` may be used to write a single JSON stream. Instances of this class are
+ * not thread safe. Calls that would result in a malformed JSON string will fail with an [IllegalStateException].
+ */
+public abstract class JsonWriter internal constructor() : Closeable, Flushable {
+  /**
+   * The nesting stack. Using a manual array rather than an ArrayList saves 20%. This stack will
+   * grow itself up to 256 levels of nesting including the top-level document. Deeper nesting is
+   * prone to trigger StackOverflowErrors.
+   */
+  @JvmField
+  internal var stackSize: Int = 0
+
+  @JvmField
+  internal var scopes: IntArray = IntArray(32)
+
+  @JvmField
+  internal var pathNames: Array<String?> = arrayOfNulls(32)
+
+  @JvmField
+  internal var pathIndices: IntArray = IntArray(32)
+
+  /**
+   * A string containing a full set of spaces for a single level of indentation, or null for no
+   * pretty printing.
+   */
+  @JvmField
+  internal var indent: String? = null
+
+  /**
+   * Configure this writer to relax its syntax rules. By default, this writer only emits well-formed
+   * JSON as specified by [RFC 7159](http://www.ietf.org/rfc/rfc7159.txt). Setting the
+   * writer to lenient permits the following:
+   *
+   * - Top-level values of any type. With strict writing, the top-level value must be an object
+   * or an array.
+   * - Numbers may be [NaNs][Double.isNaN] or [infinities][Double.isInfinite].
+   *
+   *  Returns true if this writer has relaxed syntax rules.
+   */
+  @get:CheckReturnValue
+  public var isLenient: Boolean = false
+
+  /**
+   * Sets whether object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   *
+   * Returns true if object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   */
+  @get:CheckReturnValue
+  public var serializeNulls: Boolean = false
+
+  @JvmField
+  internal var promoteValueToName: Boolean = false
+
+  /**
+   * Controls the deepest stack size that has begin/end pairs flattened:
+   *
+   * - If -1, no begin/end pairs are being suppressed.
+   * - If positive, this is the deepest stack size whose begin/end pairs are eligible to be flattened.
+   * - If negative, it is the bitwise inverse (~) of the deepest stack size whose begin/end pairs have been flattened.
+   *
+   * We differentiate between what layer would be flattened (positive) from what layer is being
+   * flattened (negative) so that we don't double-flatten.
+   *
+   * To accommodate nested flattening we require callers to track the previous state when they
+   * provide a new state. The previous state is returned from [beginFlatten] and restored
+   * with [endFlatten].
+   */
+  @JvmField
+  internal var flattenStackSize: Int = -1
+
+  private var tags: MutableMap<Class<*>, Any>? = null
+
+  /** Returns the scope on the top of the stack.  */
+  internal fun peekScope(): Int {
```

Comment:
Change these previously package-private functions to `protected` first and only use `internal` if they are accessed outside of subclasses

Reply:
48dd597e89bd532fb8542993a8a4658a410070ac